File: services/platform-core-service/app/schemas/pay_per_use/application_usage.py
"""Response schemas for the Metering Dashboard's Applications tab."""
from pydantic import BaseModel


class MoneyPercent(BaseModel):
    amount: float
    percentage: float


class ApplicationUsageSummaryResponse(BaseModel):
    totalApplications: int
    allocatedBudget: MoneyPercent
    spendBudget: MoneyPercent
    remainingBudget: MoneyPercent
    # Unlike UsageSummaryResponse.billingPeriod (a real YYYY-MM month), this is
    # always the constant "lifetime": budget_usage carries no billing_month
    # column, so every figure here is a lifetime-cumulative total, not scoped
    # to any period. There is no billing_period query param on these endpoints
    # — a dashboard period selector must not compare this figure against a
    # period-scoped one (e.g. the Overview tab's totalSpend) without
    # accounting for that difference.
    billingPeriod: str = "lifetime"


class ApplicationUsageListItem(BaseModel):
    applicationId: int
    name: str
    domain: str | None = None
    # allocatedBudget.percentage is % of the institution's total budget.
    # spendBudget/remainingBudget.percentage are % of this application's OWN allocation.
    allocatedBudget: MoneyPercent
    spendBudget: MoneyPercent
    remainingBudget: MoneyPercent


class ApplicationUsageListResponse(BaseModel):
    data: list[ApplicationUsageListItem]
    total: int


class ApiKeyUsageItem(BaseModel):
    keyId: int
    keyName: str
    maskedKey: str
    isActive: bool
    # allocatedBudget.percentage is % of the institution's total budget, computed
    # from the amount — NOT the raw api_key.allocated_percentage column, which is
    # stored as % of the parent APPLICATION's budget instead (see
    # application_usage_service.py's get_application_detail for why the two must
    # not be conflated under this one field).
    # spendBudget/remainingBudget.percentage are % of this key's OWN allocation.
    allocatedBudget: MoneyPercent
    spendBudget: MoneyPercent
    remainingBudget: MoneyPercent


class ApplicationUsageTotals(BaseModel):
    allocatedBudget: float
    spendBudget: float
    remainingBudget: float


class ApplicationUsageDetailResponse(BaseModel):
    applicationId: int
    applicationName: str
    domain: str | None = None
    allocatedBudget: MoneyPercent
    spendBudget: MoneyPercent
    remainingBudget: MoneyPercent
    apiKeys: list[ApiKeyUsageItem]
    totals: ApplicationUsageTotals

File: frontend/simple-ui/src/types/applicationUsage.ts
/** Metering Dashboard — Application Usage (US6 / AI4IDS-2894). Lifetime-cumulative figures. */

export interface MoneyPercent {
  amount: number;
  percentage: number;
}

export interface ApplicationUsageSummary {
  totalApplications: number;
  allocatedBudget: MoneyPercent;
  spendBudget: MoneyPercent;
  remainingBudget: MoneyPercent;
  billingPeriod: string;
}

export interface ApplicationUsageListItem {
  applicationId: number;
  name: string;
  domain: string | null;
  allocatedBudget: MoneyPercent;
  spendBudget: MoneyPercent;
  remainingBudget: MoneyPercent;
}

export interface ApplicationUsageListResponse {
  data: ApplicationUsageListItem[];
  total: number;
}

export interface ApiKeyUsageItem {
  keyId: number;
  keyName: string;
  maskedKey: string;
  isActive: boolean;
  allocatedBudget: MoneyPercent;
  spendBudget: MoneyPercent;
  remainingBudget: MoneyPercent;
}

export interface ApplicationUsageTotals {
  allocatedBudget: number;
  spendBudget: number;
  remainingBudget: number;
}

export interface ApplicationUsageDetail {
  applicationId: number;
  applicationName: string;
  domain: string | null;
  allocatedBudget: MoneyPercent;
  spendBudget: MoneyPercent;
  remainingBudget: MoneyPercent;
  apiKeys: ApiKeyUsageItem[];
  totals: ApplicationUsageTotals;
}

export interface ApplicationUsageListParams {
  tenantId: string;
  sortOrder?: "asc" | "desc";
  limit?: number;
  offset?: number;
}
